Covid-19 vaccine forecasting story

jack89roberts opened the pull request:
### Summary

Forecasting when the UK will fully vaccinate the adult population with data collected from the API of the government's Covid-19 dashboard, and using fairly simple approaches such as "when will the population be vaccinated if doses continue to be given at the same rate as the last week?"

Fixes #129 

### List of changes proposed in this PR (pull-request)

- New story notebook (`stories/covid-vax/covid_vax.ipynb`)
- `environment.yml` conda environment for dependencies (which are minimal, only requests, pandas, numpy, matplotlib and seaborn) - not sure how dependencies are handled for the stories?
- Backup data file `20210504_data.json` in case the API changes/is unavailable in the future.

### What should a reviewer concentrate their feedback on?

- [x] Length and level of detail ok? I feel like it got quite long for something that's actually pretty simple. E.g. is the amount of time describing the code ok - I was trying to make it as accessible as possible but might have gone over the top in places.
- [x] Approach reasonable? In particular the forecast with the uncertainty band? It's maybe not the most robust approach but I was trying to avoid adding the complexity of using an "actual" forecasting algorithm.
- [x] Is the explanation of how the forecasts decide whether doses will be given as first or second doses clear? And of calculating the delay between 1st and 2nd doses?
- [x] Does the code make sense?
- [x] Everything looks ok?
- [x] Cut off date for the data is currently set to 4th May. This could be updated closer to the time of publishing, but it's also true that the forecasts become less interesting as time goes on and the vaccine programme gets closer to completion.

### Acknowledging contributors

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [x] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file: <!-- replace this text with the GitHub IDs of any new contributors -->

jack89roberts committed:
add Covid-19 vaccine forecasting story

jack89roberts committed:
change forecast end dates to October rather than end of year

jack89roberts committed:
update binder link

jack89roberts committed:
add author

jack89roberts committed:
add affiliation

crangelsmith commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1527 @@
+{
```
<p>test</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

crangelsmith replied:
<p>It would be good at the beginning to use the full name United Kingdom (UK) and then we can use the abbreviation. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

crangelsmith replied:
<p>In other to keep the two narratives, I propose to rewrite this parragraph a bit. One saying what the figure shows ("history of the total number of people vaccinated with first and second doses"), and another that says that the plot_cumulative_doses function to builds the figure. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

crangelsmith replied:
<p>Two narratives here. Break the second paragraph in two, one talking about what the graph is going to show, and another about the function to build the figure is doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

crangelsmith replied:
<p>Two narratives, break the second paragraph after ("the number of doses delivered on each weekday".)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

crangelsmith replied:
<p>Two narratives. Break second paragraph in two after "... the total number of first doses (<code>cumFirst</code>) on a given date".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

crangelsmith replied:
<p>Two narratives. Break second paragraph in two after "of the uncertainty in the completion dates".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

crangelsmith replied:
<p>Two narratives, break the paragraph in two after "...that vaccines are given at the same rate as the year so far)."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

crangelsmith commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1527 @@
+{
```
<p>In here there are two narratives, all above "Step 1" in high level, after is technical. </p><p> </p><p>Don't worry about this point @<a href="https://github.com/jack89roberts" target="_blank">jack89roberts</a>, is a comment for the publishing step. We want to have two colours in the text, so those not very technically incline can just read the high level story and skip the technical. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

jack89roberts replied:
<p>Ok great, maybe we should add a sentence in the intro to explain this in the published version?</p>

billfinnegan replied:
<p>You mention that the vaccination data is only available as summary total for the whole UK. It looks like the dashboard also breaks the vaccinations down by the 4 nations of the UK. Is it worth mentioning that, even if the story doesn't get into those regional/devolved differences?</p>

billfinnegan replied:
<p>On the split between high level and technical, I might start the technical block with the previous paragraph: "The documentation for the API..."</p>

jack89roberts replied:
<p>Good spot for the nations data! I wonder if that's been added since I started writing this, or maybe I just missed it before.</p>

crangelsmith commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1527 @@
+{
```
<p>Two narratives. I propose to break the paragraph at "historical data first." The next sentence ("Let's take a look at the first row of our data:") can go in a new line with a technical color. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

billfinnegan replied:
<p>There is a slight inconsistency on section headers, in terms of capitalisation and the use of questions. I like the two headers that follow, so wonder if that should be a model for this and the ones above?</p>

crangelsmith commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1527 @@
+{
```
<p>Two narratives, technical side in the last paragraph (this is just a note for when we publish to change the color of it).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

billfinnegan replied:
<p>The link for update in over 50s being close to 95% relates to my concerns about how full vaccination is presented above. Wondering if this should be introduced into the high level narrative earlier.</p>

crangelsmith commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1527 @@
+{
```
<p>Thank you for this great story Jack, it reads supper well.</p><p> </p><p>I think the conclusion is missing a discussion (just a paragraph after the first one) about the limitations of these forecast (e.g model limitations, assumptions that might not be completely true, or that might change in time, etc). You did this a bit at the end of the "Alternative forecast" section but i think it needs a general overview.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

billfinnegan replied:
<p>Agreed - great story, really interesting analysis, clear explanation, super timely.</p><p> </p><p>Maybe part of this discussion is exploring what contributes to the vaccination rate: supply (doses acquired by the UK gov from the vaccine makers), demand (people getting the vaccine when offered), and process (roll-out, vaccination centres, appointments). I'm still mostly concerned about statements that imply vaccination has been completed with certain groups when there might still be some hesitancy/resistance in these groups.</p>

billfinnegan replied:
<p>Also, I just made some copyedits through Github - hope that was the right approach. Not sure what to do next. https://github.com/alan-turing-institute/TuringDataStories/compare/covid-vax-forecast...billfinnegan:patch-1</p>

jack89roberts replied:
Thanks @billfinnegan ! I'll merge in those edits and then start working through the rest of your comments 🙂

crangelsmith commented:
@all-contributors please add @jack89roberts for blog

crangelsmith commented:
@all-contributors please add @jack89roberts for idea

crangelsmith commented:
@all-contributors please add @billfinnegan for review

crangelsmith commented on stories/covid-vax/README.md:
````diff
@@ -0,0 +1,23 @@
+  ```
+  conda activate covid-vax
+  ```
+
````
How do you connect your conda environment with the jupyter notebook kernel? For me it doesn't happen by default (the default would still be Python 3, that doesn't even have pandas), so i had to use ipykernel in the way:

```
python -m ipykernel install --user --name=covid-vax
```

and then I select the kernel under that name. Is there another way? Let's make sure it's explicit.

jack89roberts replied:
Weird, for me just running `jupyter-notebook` after `conda activate covid-vax` is enough to start jupyter with it defaulting to using the right python environment. Does `which jupyter-notebook` point to the `jupyter` installed in the virtual environment for you?

Is there a more reliable way to set it up? Very happy to change it if so (and probably should change some other things I have set up the same way if this doesn't always work 😀 !)

crangelsmith replied:
Yeah, it points to the ```jupyter``` from the conda environment, . But it is strange, I have always had to do the ipykernel trick in all my notebooks and environments...

We can just leave it as it is, but add a small comment to the README saying to use the ipykernel trick if their notebooks are not directly associated to the conda environment? Maybe add this [link](https://gdcoder.com/how-to-create-and-add-a-conda-environment-as-jupyter-kernel/) as reference?

jack89roberts replied:
P.S. for me running `jupyter kernelspec list` from the environment has the default `python3` kernel set to the virtual env:
```shell
jupyter kernelspec list
Available kernels:
  # ... other kernels, then ... 
  python3     /Users/jroberts/opt/anaconda3/envs/covid-vax/share/jupyter/kernels/python3
```

crangelsmith replied:
Mine is a veeeery long list hehehe. Maybe is that if there is no other kernels defined the default goes to current conda environment?

jack89roberts committed:
define UK abbreviation

jack89roberts committed:
implement 2 narrative changes + add limitations to conclusions

jack89roberts committed:
format code cells with black

jack89roberts committed:
add help for possible conda/Jupyter kernel issues

jack89roberts committed:
tidy up code for SAGE forecasts plot

jack89roberts committed:
Merge branch 'covid-vax-forecast' of https://github.com/alan-turing-institute/TuringDataStories into covid-vax-forecast

billfinnegan commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1597 @@
+{
```
<p>I may be reading too much into a single word, but would it be better to say "Step 2: Defining which data we wish to retrieve"? Statistics implied to me some analysis has already been done on these numbers, but they are just a vaccine dose counts. Full disclosure: I'm a social scientists, not a data scientist, so I may be totally wrong on this terminology.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

jack89roberts replied:
<p>"Data"  is clear and makes the sentence shorter and snappier so let's go with that! </p>

billfinnegan commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1597 @@
+{
```
<p>I think the info on the UK pop and the priority groups should be in the high level, not technical, narrative, so we might need to think about how the high level flows and if this needs to be moved around at all (if you don't read the technical stuff looking at the first and last row, you are missing some context on this data).</p><p> </p><p>But I also wanted to follow up on one statement above: "we can see that first doses for groups 1-9 have been completed as well as second doses for groups 1-4." We can see that the # of people = to the combined pop of those groups has been vaccinated, but do we know that those groups have reached 100% of the first and second doses? Looks like there is some vaccine hesitancy in some of these priority groups, so might be worth cross-referencing https://www.england.nhs.uk/statistics/statistical-work-areas/covid-19-vaccinations/. Hopefully this isn't a can of worms - maybe you just need a caveat, and to link to this other analysis.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

jack89roberts replied:
Good point about the priority group "completion" dates. I've added a paragraph to explain the caveats with this and weakened statements about populations being vaccinated (vs. the sub-totals being helpful indicators of how much progress has been made so far).

jack89roberts replied:
<p>I've also reworked the ordering here so interacting with the data comes first followed by the description, everything before the paragraph starting "Although the first Covid-19 vaccine..." can be in  the technical narrative, then high-level narrative until the line starting "Bearing all the above in mind,..."</p>

billfinnegan commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1597 @@
+{
```
<p>Relates to the comment above - might need to reword "date a group was fully vaccinated" if there is a lingering percent of each group that has been offered the vaccine but not yet had it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

billfinnegan commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1597 @@
+{
```
<p>Does the legend for this plot need to be updated to show the fast and slow gov rate forecasts as dotted lines?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='74'/>

billfinnegan replied:
<p>Also, wondering if there might be some value in having a visualisation that shows the cumulative totals as a % of the total adult population - this could mean simply changing the label on the y-axis, but I wonder if thinking about the % would lead to any other types of visualisations. Maybe it is better to keep the y-axis consistent to easily compare the plots, but I was thinking the % vaccinated (1st dose and 2nd dose) puts that 100% date in some context, in terms of where we are now...</p>

jack89roberts replied:
<p>Ah I missed this one yesterday, I might see if I can get it to work with a second y-axis on the right-hand side showing the percentage.</p>

billfinnegan commented on stories/covid-vax/covid_vax.ipynb:
```diff
@@ -0,0 +1,1597 @@
+{
```
<p>Not sure if this is worth exploring, but any interest in an increased supply scenario? For example, the increase in doses that started mid-March (aside from the Easter dip) seems to be a higher level compared to the average from the beginning of the year to mid-March. What if the projected rate was based on the average of the previous three months?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/138/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='75'/>

jack89roberts replied:
<p>I'm reluctant to make the story longer by adding another forecast, but I've gone back and forward a lot about modifying one of the pre-existing ones. In the end I realised using the week up to 4th May includes a bank holiday, so I've changed the first forecast to use the average from the last 2 weeks to be more representative of the current situation (which changes the end date of the first forecast to 26th August instead of 31st August). But I also think it's likely that rates will ultimately fall near the end of the programme, once the remaining population that hasn't been vaccinated consists mostly of people who are at low risk from Covid-19 (i.e. young people) or are at risk but reluctant to be vaccinated. I've added a sentence or two about that as well.</p>

billfinnegan committed:
Update covid_vax.ipynb

Just a few copy edits I've done directly. I hope this is the right way to do this.

jack89roberts committed:
Merge pull request #143 from billfinnegan/patch-1

Update covid_vax.ipynb

jack89roberts committed:
implement further review comments

Mostly wording around "completion dates" for vaccination groups

jack89roberts committed:
add markers showing actual status on 12th May

jack89roberts committed:
update data dump

jack89roberts committed:
change first forecast to last 2 weeks

jack89roberts commented:
@crangelsmith @billfinnegan Thanks again for all your comments! I believe I've now implemented them all, but will give it another read-through tomorrow to double-check. This includes adding markers to show the actual latest values from today in the last plot @crangelsmith , which could be updated again on the date it will be published if you like. I've replied to some of the other suggestions separately above.

I'm happy to help with any of the next publishing steps too - we might find it needs a bit more tweaking to fit the "two narratives" idea, for example.

jack89roberts committed:
removal of API reference in non-technical narrative

jack89roberts committed:
add secondary y-axis showing % of people vaccinated

jack89roberts committed:
last read through

jack89roberts commented:
Hi again @crangelsmith @billfinnegan - I've had a final read through and made some last edits. Let me know if you have any comments on the changes, but otherwise I think it's ready 😄

crangelsmith commented:
This is fantastic @jack89roberts , i'll try to give it a second read today. Do you think you could join our meeting on Wednesday? We could work together with @samvanstroud in getting it ready for publishing (e.g different colours for the two narratives, and making sure everything looks good in html, etc).

jack89roberts commented:
Sure! There is a meeting about the RDS course 13:00-13:30, but I'm available Weds afternoon otherwise.

jack89roberts committed:
update comparison to actual data to 17th May

jack89roberts committed:
update json to 17th May

jack89roberts committed:
correct date to 16th May

samvanstroud commented:
Thanks all for writing and reviewing this story @jack89roberts @billfinnegan @crangelsmith ! Merging this now.
